### ISSUE TITLE
Nginx --devel 1.9.5 build failed in 10.11

### ISSUE THREAD

**Issue by guoxiao** [Homebrew/homebrew#44298]
The errors are as following:
```
clang: error: argument unused during compilation: '-isystem /usr/local/include'
clang: error: argument unused during compilation: '-isystem /usr/include/libxml2'
clang: error: argument unused during compilation: '-isystem /System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers'
clang: error: argument unused during compilation: '-I /usr/local/opt/openssl/include'
```

more details: https://gist.github.com/anonymous/58cd6f52f268c0306abc

**Comment by DomT4** [comment]
Those lines are being added by superenv. I'm not sure why they would be fatal errors though, and I can't reproduce locally (Having done about 15 builds earlier to fix another issue :smile_cat:). The only obvious difference between our [builds](https://gist.github.com/anonymous/e431889082c15ded2b6f) is that I have `xcode` installed.

**Comment by guoxiao** [comment]
I found a workaround by adding `ENV.append_to_cflags "-Wno-error"` to the formula and make clang report these as warnings.

**Comment by tdsmith** [comment]
The details are missing from the gist you posted. It seems like there's a -W flag we should be filtering out in superenv?

**Comment by tdsmith** [comment]
I can't reproduce this; superenv is correctly filtering `-Werror`.

**Comment by DomT4** [comment]
Yeah. Same situation as Tim in the gist I posted a couple days back. Superenv is doing the correct filtering here. Were you building under `--env=std` or anything?

**Comment by guoxiao** [comment]
I created the gist log by `brew gist-logs nginx --devel`, and don't know why these errors are not included in the gist.

I created a new gist just one minute ago. See [here](https://gist.github.com/anonymous/679810292f144ae76885).

**Comment by DomT4** [comment]
Your gist logs have lines like:

```
clang called with: -pipe -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Wno-deprecated-declarations -Werror -g -c -o objs/src/core/ngx_list.o /private/tmp/nginx20150926-37502-1nz1cdx/nginx-1.9.5/.brew_home/.ccache/tmp/ngx_list.stdout.skydiver.39350.slewTb.i
superenv added:    -isystem/usr/local/include -isystem/usr/include/libxml2 -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/openssl/include
superenv executed: clang -pipe -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Wno-deprecated-declarations -Werror -g -c -o objs/src/core/ngx_list.o /private/tmp/nginx20150926-37502-1nz1cdx/nginx-1.9.5/.brew_home/.ccache/tmp/ngx_list.stdout.skydiver.39350.slewTb.i -isystem/usr/local/include -isystem/usr/include/libxml2 -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/openssl/include
```

Which in my compile was something more like:
```
clang called with: -c -pipe -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Wno-deprecated-declarations -Werror -g -I/usr/local/Cellar/pcre/8.37/include -I/usr/local/Cellar/openssl/1.0.2d_1/include -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -o objs/src/core/ngx_file.o src/core/ngx_file.c
superenv removed:  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Werror -g -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs
superenv added:    -w -Os -march=native -Isrc/core -Isrc/event -Isrc/event/modules -Isrc/os/unix -Iobjs -isystem/usr/local/include -isystem/usr/include/libxml2 -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/openssl/include
superenv executed: clang -pipe -w -Os -march=native -c -pipe -Wno-unused-parameter -Wno-deprecated-declarations -I/usr/local/Cellar/pcre/8.37/include -I/usr/local/Cellar/openssl/1.0.2d_1/include -Isrc/core -Isrc/event -Isrc/event/modules -Isrc/os/unix -Iobjs -o objs/src/core/ngx_file.o src/core/ngx_file.c -isystem/usr/local/include -isystem/usr/include/libxml2 -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/openssl/include
```

You have no `superenv removed` lines in your logs at all :confused:.

**Comment by tdsmith** [comment]
I wonder if this is a string comparison issue!? What's `env | grep -e LANG* -e LC_*`?

**Comment by guoxiao** [comment]
@tdsmith My locale is as following:
```
LC_CTYPE=UTF-8
LANG=zh_CN.UTF-8
LC_ALL=zh_CN.UTF-8
```
However, I tried removing these settings before `brew install nginx --devel` and it still fails. So I don't think it's the reason of this issue.

**Comment by xu-cheng** [comment]
I think this may be a bug of Homebrew. Since nginx states `env :userpaths`, Homebrew will make `HOMEBREW_PREFIX/bin` as first in PATH(see [implementation code here][2]), which contributes to [the PATH used in your build][1]. As a result, it breaks the superenv. In your case, if there is a `make` in `/usr/local/bin`, it will be priority over the `Library/ENV/4.3/make` and bypass the arguments filter function. 

This behavior can cause more problems if more binaries in `/usr/local/bin` overwrite superenv. To fix it, we should make sure superenv bin comes first even for `env :userpaths`. 

[1]: https://gist.github.com/anonymous/679810292f144ae76885#file-02-make-L115
[2]: https://github.com/Homebrew/homebrew/blob/f5cc3c03fd15001a30208c04b670fa3b360f3b29/Library/Homebrew/extend/ENV/shared.rb#L196

**Comment by tdsmith** [comment]
That seems reasonable but the .cc logs show that the clang calls _are_ being intercepted and rewritten but that -Werror isn't being dropped; is that consistent?

**Comment by xu-cheng** [comment]
Yes, it's consistent. If the clang in superenv is picked up, the cc calls will be intercepted. However, the switch for refurbish args is in [`Library/ENV/4.3/make`](https://github.com/Homebrew/homebrew/blob/master/Library/ENV/4.3/make#L2), which isn't picked up in this case. 

The root problem is the build uses partial binaries in superenv, and other part in user path.

**Comment by guoxiao** [comment]
I reexamined the logs and found the reason is that I used ccache. After removing ccache this issue is gone.

**Comment by tdsmith** [comment]
Thanks for clarifying, xucheng; that makes sense!

**Comment by xu-cheng** [comment]
Should be fixed in 192b39ef9186fee8d18d7ca220aaaa5c2e435455
